Remove dependence on RefineLB

rbuch opened the pull request:
The Charm++ LB refactor has removed the CentralLB version of RefineLB,
this change removes ChaNGa's dependence on it by implementing RefineLB
directly in HierarchOrbLB. This is a stopgap change, eventually
HierarchOrbLB (and the other ChaNGa load balancers) will be rewritten
to use Charm++'s new TreeLB structure, which should make this code
considerably cleaner and more performant.

trquinn commented:
Some ChaNGa users use RefineLB.  How does one use RefineLB now?

rbuch commented:
> Some ChaNGa users use RefineLB. How does one use RefineLB now?

RefineLB can be still be used the same way it used to be, it has merely been reimplemented using the new TreeLB framework. So the same call (`+balancer RefineLB`) will call into the new version.

trquinn commented:
If I run "ChaNGa +balancer RefineLB", I get:
```
./ChaNGa +balancer RefineLB
Charm++> No provisioning arguments specified. Running with a single PE.
         Use +auto-provision to fully subscribe resources or +p1 to silence this message.
Charm++: standalone mode (not using charmrun)
Charm++> Running in SMP mode: 1 processes, 1 worker threads (PEs) + 1 comm threads per process, 1 PEs total
Charm++> The comm. thread both sends and receives messages
Converse/Charm++ Commit ID: v6.10.1-0-gcc60a79
Warning> Randomization of virtual memory (ASLR) is turned on in the kernel, thread migration may not work! Run 'echo 0 > /proc/sys/kernel/randomize_va_space' as root to disable it, or try running with '+isomalloc_sync'.
Charm++> scheduler running in netpoll mode.
CharmLB> Load balancer assumes all CPUs are same.
Charm++> Running on 1 hosts (1 sockets x 4 cores x 2 PUs = 8-way SMP)
Charm++> cpu topology info is gathered in 0.004 seconds.
Abort: Unknown load balancer: 'RefineLB'!

Available load balancers:
* CentralLB:    CentralLB base class
* MultistepLB_notopo:   Works best with multistepped runs; uses Orb3D_notopo
* MultistepLB:  Works best with multistepped runs; uses Orb3D for larger steps, greedy otherwise
* MultistepNodeLB_notopo:       Works best with multistepped runs; uses Orb3D_notopo for larger steps, greedy otherwise
* Orb3dLB:      3d ORB mapping of tree piece space onto 3d processor mesh
* Orb3dLB_notopo:       3d ORB mapping of tree piece space onto 3d processor mesh
* HybridBaseLB: HybridBase load balancer
* HierarchOrbLB:        Hybrid load balancer

------------- Processor 0 Exiting: Called CmiAbort ------------
Reason: Abort
[0] Stack Traceback:
  [0:0] ChaNGa 0x7e5d56 CmiAbortHelper(char const*, char const*, char const*, int, int)
  [0:1] ChaNGa 0x7e5e6d 
  [0:2] ChaNGa 0x76db4c 
  [0:3] ChaNGa 0x770586 LBDBInit::LBDBInit(CkArgMsg*)
```

rbuch commented:
> If I run "ChaNGa +balancer RefineLB", I get:
> 
> ```
> ./ChaNGa +balancer RefineLB
> Charm++> No provisioning arguments specified. Running with a single PE.
>          Use +auto-provision to fully subscribe resources or +p1 to silence this message.
> Charm++: standalone mode (not using charmrun)
> Charm++> Running in SMP mode: 1 processes, 1 worker threads (PEs) + 1 comm threads per process, 1 PEs total
> Charm++> The comm. thread both sends and receives messages
> Converse/Charm++ Commit ID: v6.10.1-0-gcc60a79
> Warning> Randomization of virtual memory (ASLR) is turned on in the kernel, thread migration may not work! Run 'echo 0 > /proc/sys/kernel/randomize_va_space' as root to disable it, or try running with '+isomalloc_sync'.
> Charm++> scheduler running in netpoll mode.
> CharmLB> Load balancer assumes all CPUs are same.
> Charm++> Running on 1 hosts (1 sockets x 4 cores x 2 PUs = 8-way SMP)
> Charm++> cpu topology info is gathered in 0.004 seconds.
> Abort: Unknown load balancer: 'RefineLB'!
> 
> Available load balancers:
> * CentralLB:    CentralLB base class
> * MultistepLB_notopo:   Works best with multistepped runs; uses Orb3D_notopo
> * MultistepLB:  Works best with multistepped runs; uses Orb3D for larger steps, greedy otherwise
> * MultistepNodeLB_notopo:       Works best with multistepped runs; uses Orb3D_notopo for larger steps, greedy otherwise
> * Orb3dLB:      3d ORB mapping of tree piece space onto 3d processor mesh
> * Orb3dLB_notopo:       3d ORB mapping of tree piece space onto 3d processor mesh
> * HybridBaseLB: HybridBase load balancer
> * HierarchOrbLB:        Hybrid load balancer
> 
> ------------- Processor 0 Exiting: Called CmiAbort ------------
> Reason: Abort
> [0] Stack Traceback:
>   [0:0] ChaNGa 0x7e5d56 CmiAbortHelper(char const*, char const*, char const*, int, int)
>   [0:1] ChaNGa 0x7e5e6d 
>   [0:2] ChaNGa 0x76db4c 
>   [0:3] ChaNGa 0x770586 LBDBInit::LBDBInit(CkArgMsg*)
> ```

Aha, I had removed the explicit linking of Refine, Greedy, and Rotate in `Makefile.in`. I've added the TreeLB module in ae53a59ab3a89e2a729239241ae3793c83194d5e, which now will provide the functionality of the aforementioned LBs (plus a couple more). Now it should work with `+balancer RefineLB`.

rbuch commented:
> > If I run "ChaNGa +balancer RefineLB", I get:
> > ```
> > ./ChaNGa +balancer RefineLB
> > Charm++> No provisioning arguments specified. Running with a single PE.
> >          Use +auto-provision to fully subscribe resources or +p1 to silence this message.
> > Charm++: standalone mode (not using charmrun)
> > Charm++> Running in SMP mode: 1 processes, 1 worker threads (PEs) + 1 comm threads per process, 1 PEs total
> > Charm++> The comm. thread both sends and receives messages
> > Converse/Charm++ Commit ID: v6.10.1-0-gcc60a79
> > Warning> Randomization of virtual memory (ASLR) is turned on in the kernel, thread migration may not work! Run 'echo 0 > /proc/sys/kernel/randomize_va_space' as root to disable it, or try running with '+isomalloc_sync'.
> > Charm++> scheduler running in netpoll mode.
> > CharmLB> Load balancer assumes all CPUs are same.
> > Charm++> Running on 1 hosts (1 sockets x 4 cores x 2 PUs = 8-way SMP)
> > Charm++> cpu topology info is gathered in 0.004 seconds.
> > Abort: Unknown load balancer: 'RefineLB'!
> > 
> > Available load balancers:
> > * CentralLB:    CentralLB base class
> > * MultistepLB_notopo:   Works best with multistepped runs; uses Orb3D_notopo
> > * MultistepLB:  Works best with multistepped runs; uses Orb3D for larger steps, greedy otherwise
> > * MultistepNodeLB_notopo:       Works best with multistepped runs; uses Orb3D_notopo for larger steps, greedy otherwise
> > * Orb3dLB:      3d ORB mapping of tree piece space onto 3d processor mesh
> > * Orb3dLB_notopo:       3d ORB mapping of tree piece space onto 3d processor mesh
> > * HybridBaseLB: HybridBase load balancer
> > * HierarchOrbLB:        Hybrid load balancer
> > 
> > ------------- Processor 0 Exiting: Called CmiAbort ------------
> > Reason: Abort
> > [0] Stack Traceback:
> >   [0:0] ChaNGa 0x7e5d56 CmiAbortHelper(char const*, char const*, char const*, int, int)
> >   [0:1] ChaNGa 0x7e5e6d 
> >   [0:2] ChaNGa 0x76db4c 
> >   [0:3] ChaNGa 0x770586 LBDBInit::LBDBInit(CkArgMsg*)
> > ```
> 
> Aha, I had removed the explicit linking of Refine, Greedy, and Rotate in `Makefile.in`. I've added the TreeLB module in [ae53a59](https://github.com/N-BodyShop/changa/commit/ae53a59ab3a89e2a729239241ae3793c83194d5e), which now will provide the functionality of the aforementioned LBs (plus a couple more). Now it should work with `+balancer RefineLB`.

This was breaking in older versions of Charm++, I've now changed the implementation in Charm++ to accept legacy LB names as a parameter to `-module` (previously it only worked with `-balancer`), so now this should work with both pre and post TreeLB versions of Charm++.

rbuch commented:
@trquinn Just a ping to see if you can take a look and merge this. It should all be good to go.
